packages/playwright-core/src/server/registry/download.ts

```diff
@@ -15,68 +15,38 @@
 
-type OnProgressCallback = (downloadedBytes: number, totalBytes: number) => void;
 type DownloadFileLogger = (message: string) => void;
 type DownloadFileOptions = {
-  progressCallback?: OnProgressCallback,
+  progressBarName?: string,
   log?: DownloadFileLogger,
   userAgent?: string
 };
 
-function downloadFile(url: string, destinationPath: string, options: DownloadFileOptions = {}): Promise<{ error: any }> {
-  const {
-    progressCallback,
-    log = () => { },
-  } = options;
-  log(`running download:`);
-  log(`-- from url: ${url}`);
-  log(`-- to location: ${destinationPath}`);
-  let fulfill: ({ error }: { error: any }) => void = ({ error }) => { };
-  let downloadedBytes = 0;
-  let totalBytes = 0;
-
-  const promise: Promise<{ error: any }> = new Promise(x => { fulfill = x; });
-
-  httpRequest({
-    url,
-    headers: options.userAgent ? {
-      'User-Agent': options.userAgent,
-    } : undefined,
-    timeout: 10_000,
-  }, response => {
-    log(`-- response status code: ${response.statusCode}`);
-    if (response.statusCode !== 200) {
-      let content = '';
-      const handleError = () => {
-        const error = new Error(`Download failed: server returned code ${response.statusCode} body '${content}'. URL: ${url}`);
-        // consume response data to free up memory
-        response.resume();
-        fulfill({ error });
-      };
-      response
-          .on('data', chunk => content += chunk)
-          .on('end', handleError)
-          .on('error', handleError);
+function downloadFileOutOfProcess(url: string, destinationPath: string, options: DownloadFileOptions = {}): Promise<{ error: Error | null }> {
+  const cp = childProcess.fork(path.join(__dirname, 'downloadMain.js'), [url, destinationPath, options.progressBarName || '', options.userAgent || '']);
+  const promise = new ManualPromise<{ error: Error | null }>();
+  cp.on('message', (message: any) => {
+    if (message?.method === 'log')
+      options.log?.(message.params.message);
+  });
+  cp.on('exit', code => {
+    if (code !== 0) {
+      promise.resolve({ error: new Error(`Download process exited with code ${code}`) });
       return;
     }
-    const file = fs.createWriteStream(destinationPath);
-    file.on('finish', () => fulfill({ error: null }));
-    file.on('error', error => fulfill({ error }));
-    response.pipe(file);
-    totalBytes = parseInt(response.headers['content-length'] || '0', 10);
-    log(`-- total bytes: ${totalBytes}`);
-    if (progressCallback)
-      response.on('data', onData);
-  }, (error: any) => fulfill({ error }));
+    // Node.js has a bug where the process can exit with 0 code even though there was an uncaught exception.
+    // Thats why we manually check if the destination file exists.
+    // https://github.com/microsoft/playwright/issues/17394
+    if (!fs.existsSync(destinationPath))
+      promise.resolve({ error: new Error(`Download process failed to write to ${destinationPath}`) });
```

Comment:
Will this trigger a retry?

Reply:
Yes, see https://github.com/mxschmitt/playwright/blob/5284992135560cb7b0edcbf7f225a05d49b943a4/packages/playwright-core/src/server/registry/download.ts#L78